DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/svd/src/svd_testbench_tool.hpp

```diff
@@ -0,0 +1,167 @@
+#ifndef __SVD_TESTBENCH_TOOL_HPP__
+#define __SVD_TESTBENCH_TOOL_HPP__
+
+#include <iostream>
+#include <vector>
+
+/*
+ * This file contains helper functions that is used in the host test bench
+ * of the SVD demonstration to check for correctness.
+ */
+
+#define EPSILON 2E-6
+
+namespace svd_testbench_tool {  // not for kernel code
+
+template <typename TT>
+void SoftTranspose(std::vector<std::vector<TT>> &origin, unsigned rows,
+                   unsigned cols, std::vector<std::vector<TT>> &transposed) {
+  // just swap row and col
+  for (unsigned row = 0; row < rows; row++) {
+    for (unsigned col = 0; col < cols; col++) {
+      transposed[col][row] = origin[row][col];
+    }
+  }
+}
+
+// for col major matrix
+template <typename TT>
+void SoftTranspose(std::vector<TT> &mat_A, unsigned rows, unsigned cols,
+                   std::vector<TT> &mat_At) {
+  for (int i = 0; i < (rows * cols); i++) {
+    int cur_col = int(i / rows);
+    int cur_row = i % rows;
+    mat_At[cur_row * cols + cur_col] = mat_A[i];
+  }
+}
+
+template <typename TT>
+void SoftMatmult(std::vector<std::vector<TT>> &mat_A, unsigned rows_A,
+                 unsigned cols_A, std::vector<std::vector<TT>> &mat_B,
+                 unsigned rows_B, unsigned cols_B,
+                 std::vector<std::vector<TT>> &mat_AB) {
+  assert((cols_A == rows_B) && "Mat_Mult with illegal matrix sizes");
+  // Initializing AB to 0s
+  for (unsigned row = 0; row < rows_A; row++) {
+    for (unsigned col = 0; col < cols_B; col++) {
+      mat_AB[row][col] = 0.0;
+    }
+  }
+  // Multiplying matrix A and B and storing in AB.
+  for (unsigned row = 0; row < rows_A; row++) {
+    for (unsigned col = 0; col < cols_B; col++) {
+      for (unsigned item = 0; item < cols_A; item++) {
+        mat_AB[row][col] += mat_A[row][item] * mat_B[item][col];
+      }
+    }
+  }
+}
+
+// for col major matrix
+template <typename TT>
+void SoftMatmult(std::vector<TT> &mat_A, unsigned rows_A, unsigned cols_A,
+                 std::vector<TT> &mat_B, unsigned rows_B, unsigned cols_B,
+                 std::vector<TT> &mat_AB) {
+  std::vector<std::vector<TT>> a_2d(rows_A, std::vector<TT>(cols_A, 0));
+  std::vector<std::vector<TT>> b_2d(rows_B, std::vector<TT>(cols_B, 0));
+  std::vector<std::vector<TT>> ab_2d(rows_A, std::vector<TT>(cols_B, 0));
+
+  // turn A vertical
+  for (int i = 0; i < (rows_A * cols_A); i++) {
```

Comment:
You should look at consistent usage of "int" vs "unsigned" in loops.    Here both rows_A & cols_A are unsigned by you are using an "int" index variable.    You need to use the right type for the expressions used.